assets/js/contact-form.js

```diff
@@ -0,0 +1,270 @@
+    }
+  }
+
+  /** @param {string} value */
+  function validateEmail(value) {
+    if (!value) {
+      showErrorMessage(inputEmail, "Please provide a valid email address.");
+    } else if (!emailRegEx.test(value)) {
+      showErrorMessage(inputEmail, "Is your email spelled correctly?");
+    } else {
+      hideErrorMessage(inputEmail);
+    }
+  }
+
+  /** @param {string} value */
+  function validateNote(value) {
+    if (!value) {
+      showErrorMessage(
+        textAreaMessage,
+        "Please tell me what you're writing me about."
+      );
+    } else if (
+      textAreaMessage.validity.tooShort ||
+      value < MIN_MSG_LENGTH_CHARS
+    ) {
+      showErrorMessage(
+        textAreaMessage,
+        "That's an awfully short message, you better not be spamming me!"
+      );
+    } else {
+      hideErrorMessage(textAreaMessage);
+    }
+  }
+
+  /**
+   * @param {FormData} data
+   * */
+  function validateForm(data) {
+    const nameValue = data.get("name");
+    const noteValue = data.get("message");
+    const emailValue = data.get("_replyto");
+    validateName(nameValue);
+    validateNote(noteValue);
+    validateEmail(emailValue);
+  }
+
+  /** @returns {NodeListOf<HTMLElement>} */
+  function getInvalidFormFields() {
+    return form.querySelectorAll("[aria-invalid='true']");
+  }
+
+  /**
+   * - handles a client or server side form validation error
+   * @param {(HTMLInputElement | HTMLTextAreaElement)[] | string} [errors] either a string of error text that should be shown or an array of form fields that are in an invalid state.
+   */
+  function handleInvalidFormState(errors) {
+    // default error message
+    let errorMessage =
+      "Something went wrong when submitting the form. Please try submitting the form again or alternatively you may send me an email. Thanks!";
+
+    // errors from Formspree server validation
+    if (typeof errors === "string") {
+      errorMessage = errors;
+    }
+
+    // client side form validation errors
+    if (Array.isArray(errors) && errors.length) {
+      const pluralHandler = errors.length > 1 ? "s" : "";
+      const fieldTypes = errors
+        .map((element) => (element.type === "email" ? "email" : element.name))
+        .join(" and ");
+      errorMessage = `Please correct the error${pluralHandler} in the ${fieldTypes} form field${pluralHandler}.`;
+    }
+
+    // for any error situation
+    formStatus.innerText = errorMessage;
+    formStatus.classList = "error";
+    formStatus.removeAttribute("hidden");
+    formStatus.focus();
+  }
+
+  function handleFormSubmitSuccess() {
+    formStatus.classList = "success";
+    formStatus.removeAttribute("hidden");
+    formStatus.innerText =
+      "Thanks for reaching out! I'll get back to you as soon as I can.";
+    formStatus.focus();
+    form.reset();
+  }
+
+  function clearFormStatus() {
+    formStatus.classList = "";
+    formStatus.setAttribute("hidden", "");
+  }
+
+  /**
+   * {% raw %}
```

Comment:
Note: the `{% raw %}` and `{% endraw %}` comments are necessary to prevent Nunjucks from interpreting the double curly braces in the JSDoc `@typedef`s